File: phon/rust/phon-ir/src/descriptor.rs
//! The descriptor model — how an implementation reads and constructs its own
//! language's in-memory values for a given schema.
//!
//! A descriptor is a tree shaped like the schema, each node annotated with the
//! process-local facts the engine needs to read that part of a value (encode)
//! and construct it (decode). It is never transmitted, never hashed, never part
//! of schema identity. The *shape* is shared across implementations and
//! documented once in the spec; each implementation has its own descriptors
//! describing its own memory.
//!
//! Facts come in two forms: **direct facts** (offsets, strides, tags, niches)
//! the engine reads/writes itself, and **thunks** (named same-language functions)
//! for everything direct facts can't express. A node may give direct facts for
//! one direction and a thunk for the other (`r[descriptors.encode-decode-asymmetry]`).
//!
//! The generic memory-descriptor vocabulary lives in `weavy`; PHON specializes it
//! with `phon_schema::SchemaRef` so existing callers keep the same PHON-facing
//! names while other bytecode consumers can reuse the same model.
//!
//! Spec: "The descriptor model" (`r[descriptors.*]`).

use phon_schema::SchemaRef;

pub use weavy::mem::{
    ByteOwner, ByteRange, Construct, FieldDefault, Layout, MapStorage, Presence,
    RecordByteOwnership, SequenceStorage, SetStorage, Tag, Thunk,
};

/// A node of the PHON descriptor tree.
// r[impl descriptors.separate-implementations]
// r[impl descriptors.fact-driven]
pub type Descriptor = weavy::mem::Descriptor<SchemaRef>;

/// How a PHON value's bytes are read and constructed.
// r[impl descriptors.encode-decode-asymmetry]
pub type Access = weavy::mem::Access<SchemaRef>;

/// A PHON struct or tuple: its fields at offsets, with how to construct it.
pub type RecordAccess = weavy::mem::RecordAccess<SchemaRef>;

/// One PHON field: its byte offset within the record, and its descriptor.
pub type FieldAccess = weavy::mem::FieldAccess<SchemaRef>;

/// A PHON sum type: a tag selecting the active variant, and the per-variant payloads.
pub type EnumAccess = weavy::mem::EnumAccess<SchemaRef>;

/// One PHON variant: its schema index, local tag selector, and payload fields.
pub type VariantAccess = weavy::mem::VariantAccess<SchemaRef>;

/// An optional PHON value: how presence is read/written, and the some-payload.
pub type OptionAccess = weavy::mem::OptionAccess<SchemaRef>;

/// A dynamic homogeneous PHON sequence or byte sequence.
pub type SequenceAccess = weavy::mem::SequenceAccess<SchemaRef>;

/// A PHON set: its element descriptor and storage strategy.
pub type SetAccess = weavy::mem::SetAccess<SchemaRef>;

/// A PHON result: the `Ok` and `Err` payload descriptors plus thunks.
pub type ResultAccess = weavy::mem::ResultAccess<SchemaRef>;

/// A PHON owning pointer: its pointee descriptor and thunks.
pub type PointerAccess = weavy::mem::PointerAccess<SchemaRef>;

/// PHON key/value pairs: key and value descriptors plus map storage.
pub type MapAccess = weavy::mem::MapAccess<SchemaRef>;

/// A PHON runtime-shape tensor.
pub type TensorAccess = weavy::mem::TensorAccess<SchemaRef>;

File: phon/swift/phon-ir/Sources/PhonIR/Descriptor.swift
// The descriptor model — how an implementation reads and constructs its own
// language's in-memory values for a given schema.
//
// A descriptor is a tree shaped like the schema, each node annotated with the
// process-local facts the engine needs to read that part of a value (encode) and
// construct it (decode). It is never transmitted, never hashed, never part of
// schema identity. The shape is shared across implementations; each has its own
// descriptors describing its own memory.
//
// Mirrors `rust/phon-ir/src/descriptor.rs`. Where Rust names thunks (resolved to
// function pointers by a separate binding step), Swift carries the closures
// directly — Swift binds at descriptor-build time, so the name indirection is
// unnecessary.
//
// This first cut models what the typed engine handles now: scalars and in-place
// records. Enums, options, sequences, maps, result, dynamic, and opaque grow the
// `Access` enum as the engine learns them.

import PhonSchema

/// A node of the descriptor tree: the schema it realizes, its process-local
/// memory layout, and how to read and construct it.
public struct Descriptor {
    /// The schema this node realizes.
    public var schema: SchemaRef
    /// Process-local size and alignment.
    public var layout: Layout
    /// How to read and construct this value.
    public var access: Access

    public init(schema: SchemaRef, layout: Layout, access: Access) {
        self.schema = schema
        self.layout = layout
        self.access = access
    }
}

/// Process-local size and alignment, in bytes.
public struct Layout: Equatable {
    public var size: Int
    public var align: Int

    public init(size: Int, align: Int) {
        self.size = size
        self.align = align
    }
}

/// How a value's bytes are read and constructed.
public indirect enum Access {
    /// A fixed-width scalar whose in-memory bytes equal its wire bytes (bool, the
    /// integer and float primitives, char): copy `layout.size` bytes either
    /// direction. Assumes host byte order matches the wire's (little-endian).
    case scalar
    /// A struct or tuple: fields at fixed offsets.
    case record(RecordAccess)
}

/// A struct or tuple: its fields at offsets, with how to construct it.
public struct RecordAccess {
    public var fields: [FieldAccess]
    public var construct: Construct

    public init(fields: [FieldAccess], construct: Construct) {
        self.fields = fields
        self.construct = construct
    }
}

/// One field: its byte offset within the record, and its descriptor.
public struct FieldAccess {
    public var offset: Int
    public var descriptor: Descriptor
    /// How to write this field's default in place, for the decode-compat path
    /// when the field is reader-only (absent from the writer). `nil` for a
    /// required field, whose absence from the writer makes the schemas
    /// incompatible.
    public var defaultInit: ((_ slot: UnsafeMutableRawPointer) -> Void)?

    public init(
        offset: Int,
        descriptor: Descriptor,
        defaultInit: ((_ slot: UnsafeMutableRawPointer) -> Void)? = nil
    ) {
        self.offset = offset
        self.descriptor = descriptor
        self.defaultInit = defaultInit
    }
}

/// How a record is built on decode.
public enum Construct {
    /// Decode writes each field into its offset in uninitialized storage; the
    /// value is valid once all fields are written. Plain structs and tuples.
    case inPlace
    /// Decode fills a scratch buffer, then a closure builds the real value from
    /// it. Types with construction invariants, or that can't be poked field by
    /// field.
    case thunk((_ scratch: UnsafeRawPointer, _ slot: UnsafeMutableRawPointer) -> Void)
}
